src/metabase/query_processor/util/add_alias_info.clj

```diff
@@ -235,32 +235,38 @@
             (m/find-first (fn [[_field an-id-or-name _opts]]
                             (= an-id-or-name id-or-name))
                           field-exports)))
-        ;; look for a matching expression clause with the same name if still no match
+        ;; otherwise if this is a nominal field literal ref then look for matches based on the string name used
         (when-let [field-name (let [[_ id-or-name] field-clause]
                                 (when (string? id-or-name)
                                   id-or-name))]
-          (or ;; Expressions by exact name.
+          (or ;; First, look for Expressions or fields from the source query stage whose `::desired-alias` matches the
+              ;; name we're searching for.
+              (m/find-first (fn [[tag _id-or-name {::keys [desired-alias], :as _opts} :as _ref]]
+                              (when (#{:expression :field} tag)
+                                (= desired-alias field-name)))
+                            all-exports)
+              ;; Expressions by exact name.
               (m/find-first (fn [[_ expression-name :as _expression-clause]]
                               (= expression-name field-name))
                             (filter (partial mbql.u/is-clause? :expression) all-exports))
-              ;; Expressions whose ::desired-alias matches the name we're searching for.
-              (m/find-first (fn [[_expression _expression-name {::keys [desired-alias]} :as _expression-clause]]
-                              (= desired-alias field-name))
-                            (filter (partial mbql.u/is-clause? :expression) all-exports))
-              (m/find-first (fn [[_ _ opts :as _aggregation-options-clause]]
-                              (= (::source-alias opts) field-name))
-                            (filter (partial mbql.u/is-clause? :aggregation-options) all-exports))))
-        ;; look for a field referenced by the name in source-metadata
-        (let [field-name (second field-clause)]
-          (when (string? field-name)
```

Comment:
The "if this is a string field name" stuff here was duplicated with the  "if this is a string field name" stuff directly above so I combined it into one section, so this party is really just whitespace changes